hbase-server/src/test/java/org/apache/hadoop/hbase/master/assignment/TestTRSPPersistUninitializedSubProc.java

```diff
@@ -0,0 +1,125 @@
+import org.apache.hadoop.hbase.HBaseTestingUtil;
+import org.apache.hadoop.hbase.ServerName;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.RegionInfo;
+import org.apache.hadoop.hbase.master.HMaster;
+import org.apache.hadoop.hbase.master.assignment.TransitRegionStateProcedure.TransitionType;
+import org.apache.hadoop.hbase.master.procedure.MasterProcedureEnv;
+import org.apache.hadoop.hbase.master.procedure.ServerCrashProcedure;
+import org.apache.hadoop.hbase.procedure2.Procedure;
+import org.apache.hadoop.hbase.procedure2.ProcedureExecutor;
+import org.apache.hadoop.hbase.procedure2.ProcedureSuspendedException;
+import org.apache.hadoop.hbase.procedure2.ProcedureTestingUtility;
+import org.apache.hadoop.hbase.procedure2.ProcedureYieldException;
+import org.apache.hadoop.hbase.testclassification.MasterTests;
+import org.apache.hadoop.hbase.testclassification.MediumTests;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import org.apache.hadoop.hbase.shaded.protobuf.generated.ProcedureProtos.ProcedureState;
+
+/**
+ * Testcase for HBASE-29259
+ */
+@Category({ MasterTests.class, MediumTests.class })
+public class TestTRSPPersistUninitializedSubProc {
+
+  @ClassRule
+  public static final HBaseClassTestRule CLASS_RULE =
+    HBaseClassTestRule.forClass(TestTRSPPersistUninitializedSubProc.class);
+
+  private static HBaseTestingUtil UTIL = new HBaseTestingUtil();
+
+  private static byte[] CF = Bytes.toBytes("cf");
+
+  private static TableName TN = TableName.valueOf("tn");
+
+  public static class TRSPForTest extends TransitRegionStateProcedure {
+
+    private boolean injected = false;
+
+    public TRSPForTest() {
+    }
+
+    public TRSPForTest(MasterProcedureEnv env, RegionInfo hri, ServerName assignCandidate,
+      boolean forceNewPlan, TransitionType type) {
+      super(env, hri, assignCandidate, forceNewPlan, type);
+    }
+
+    @Override
+    protected Procedure[] execute(MasterProcedureEnv env)
+      throws ProcedureSuspendedException, ProcedureYieldException, InterruptedException {
+      Procedure[] subProcs = super.execute(env);
+      if (!injected && subProcs != null && subProcs[0] instanceof CloseRegionProcedure) {
+        injected = true;
+        ServerName sn = ((CloseRegionProcedure) subProcs[0]).targetServer;
+        env.getMasterServices().getServerManager().expireServer(sn);
+        try {
+          UTIL.waitFor(15000, () -> env.getMasterServices().getProcedures().stream().anyMatch(
+            p -> p instanceof ServerCrashProcedure && p.getState() != ProcedureState.INITIALIZING));
+        } catch (IOException e) {
+          throw new UncheckedIOException(e);
+        }
+        // sleep 10 seconds to let the SCP interrupt the TRSP, where we will call TRSP.serverCrashed
+        Thread.sleep(10000);
```

Comment:
Will this be long enough for consistent behavior on Jenkins? What if this thread blocked indefinitely, would the rest of the system process as expected?

Reply:
Ah this is not very stable, but at least the test will be flaky without the fixing in this PR. And since in this PR we delay the releasing of the lock, the problem just disappear, no matter how long you sleep, so here I just use a simple 10 seconds.